File: shared/components/src/MapView/array-offset.ts
/**
 * Array offset calculations for towed-array sensors.
 *
 * Three array centre modes determine where a sensor's bearing lines originate:
 *   - PLAIN: backtrack along the vessel's heading by the offset distance
 *   - WORM: walk backwards along the vessel's track path by the offset distance
 *   - MEASURED: interpolate from the sensor's measured position time-series,
 *               with fallback to PLAIN when the contact timestamp falls
 *               outside the measured range
 *
 * All calculations are pure functions with identical behaviour to the Python
 * implementation in services/calc/debrief_calc/tools/sensor/array_offset.py.
 * Parity is verified by cross-language golden tests.
 */

import type { SensorData } from '@debrief/schemas';
import { geodesicDestination, interpolateTrackCourse, interpolateTrackPosition } from './sensor-utils';
import { findNearestPointIndex } from './temporal-utils';

/** Mean Earth radius in metres (matches Python haversine_distance_metres) */
const EARTH_RADIUS_METRES = 6_371_000;

/**
 * Geodesic distance between two points in metres.
 *
 * Uses the haversine formula with the mean Earth radius (6371000m).
 *
 * @param lon1 Longitude of point 1 (degrees)
 * @param lat1 Latitude of point 1 (degrees)
 * @param lon2 Longitude of point 2 (degrees)
 * @param lat2 Latitude of point 2 (degrees)
 * @returns Distance in metres
 */
export function haversineDistanceMetres(
  lon1: number,
  lat1: number,
  lon2: number,
  lat2: number,
): number {
  const toRad = Math.PI / 180;
  const phi1 = lat1 * toRad;
  const phi2 = lat2 * toRad;
  const dPhi = (lat2 - lat1) * toRad;
  const dLambda = (lon2 - lon1) * toRad;

  const a =
    Math.sin(dPhi / 2) * Math.sin(dPhi / 2) +
    Math.cos(phi1) * Math.cos(phi2) * Math.sin(dLambda / 2) * Math.sin(dLambda / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_METRES * c;
}

/**
 * PLAIN mode: backtrack from the vessel's position along its heading.
 *
 * The reverse bearing is `(courseDeg + 180) mod 360`. The array centre is the
 * geodesic destination from the vessel at that reverse bearing, at the offset
 * distance.
 *
 * @param hostPosition Vessel position [lon, lat]
 * @param courseDeg Vessel heading (degrees, 0-360)
 * @param offsetMetres Backtrack distance in metres (must be >= 0)
 * @returns Array centre [lon, lat]
 */
export function computePlainOffset(
  hostPosition: [number, number],
  courseDeg: number,
  offsetMetres: number,
): [number, number] {
  if (offsetMetres <= 0) return hostPosition;
  const reverseBearing = ((courseDeg + 180) % 360 + 360) % 360;
  return geodesicDestination(hostPosition, reverseBearing, offsetMetres);
}

/**
 * WORM mode: walk backward along the vessel's track path.
 *
 * Starts at the interpolated vessel position at the contact time, then walks
 * backward through track segments accumulating geodesic distances until the
 * offset distance is reached. If the track is exhausted before reaching the
 * offset, the array centre is placed at the earliest available track point.
 *
 * @param trackCoordinates Track geometry coordinates [lon, lat][]
 * @param trackPositions Track positions with ISO-8601 timestamps
 * @param contactTimeMs Contact timestamp (epoch ms)
 * @param offsetMetres Distance to walk backward along the track path
 * @returns Array centre [lon, lat] on the track path
 */
export function backtrackAlongTrack(
  trackCoordinates: [number, number][],
  trackPositions: Array<{ time: string }>,
  contactTimeMs: number,
  offsetMetres: number,
): [number, number] {
  if (trackCoordinates.length === 0) {
    throw new Error('backtrackAlongTrack: empty track coordinates');
  }
  if (trackCoordinates.length === 1 || offsetMetres <= 0) {
    return trackCoordinates[0]!;
  }

  // Interpolate starting point at contact time (falls back to earliest when out of range)
  const startPoint =
    interpolateTrackPosition(trackCoordinates, trackPositions, contactTimeMs) ??
    trackCoordinates[0]!;

  // Determine starting segment index: the largest index i where positions[i].time <= contactTimeMs
  const timestamps = trackPositions.map((p) => Date.parse(p.time));
  let startIdx = 0;
  if (contactTimeMs >= timestamps[timestamps.length - 1]!) {
    startIdx = trackCoordinates.length - 1;
  } else if (contactTimeMs <= timestamps[0]!) {
    return trackCoordinates[0]!;
  } else {
    const nearest = findNearestPointIndex(timestamps, contactTimeMs);
    startIdx = timestamps[nearest]! <= contactTimeMs ? nearest : nearest - 1;
    if (startIdx < 0) startIdx = 0;
  }

  let remaining = offsetMetres;
  let currentPoint: [number, number] = [startPoint[0], startPoint[1]];

  for (let i = startIdx; i >= 1; i--) {
    const prevPoint = trackCoordinates[i - 1]!;
    const segLen = haversineDistanceMetres(
      currentPoint[0],
      currentPoint[1],
      prevPoint[0],
      prevPoint[1],
    );

    if (segLen === 0) {
      currentPoint = prevPoint;
      continue;
    }

    if (remaining <= segLen) {
      const fraction = remaining / segLen;
      return [
        currentPoint[0] + (prevPoint[0] - currentPoint[0]) * fraction,
        currentPoint[1] + (prevPoint[1] - currentPoint[1]) * fraction,
      ];
    }

    remaining -= segLen;
    currentPoint = prevPoint;
  }

  // Track exhausted: return the earliest point on the track
  return trackCoordinates[0]!;
}

/**
 * MEASURED mode: interpolate from the sensor's measured position time-series.
 *
 * If the contact time is outside the measured range, returns null so the
 * caller can fall back to PLAIN mode (FR-004).
 *
 * Measured positions are sorted by time before lookup to handle unordered input.
 *
 * @param measuredPositions Time-series of measured array positions
 * @param contactTimeMs Contact timestamp (epoch ms)
 * @returns Interpolated [lon, lat], or null if out of measured range
 */
export function interpolateMeasuredPosition(
  measuredPositions: Array<{ time: string; location: number[] }>,
  contactTimeMs: number,
): [number, number] | null {
  if (!measuredPositions || measuredPositions.length === 0) return null;

  const parsed = measuredPositions
    .map((p) => ({ t: Date.parse(p.time), lon: p.location[0]!, lat: p.location[1]! }))
    .filter((p) => !isNaN(p.t))
    .sort((a, b) => a.t - b.t);

  if (parsed.length === 0) return null;

  const first = parsed[0]!;
  const last = parsed[parsed.length - 1]!;

  if (contactTimeMs < first.t || contactTimeMs > last.t) return null;

  if (contactTimeMs === first.t) return [first.lon, first.lat];
  if (contactTimeMs === last.t) return [last.lon, last.lat];

  // Linear search for bracketing pair (arrays are typically small)
  for (let i = 0; i < parsed.length - 1; i++) {
    const a = parsed[i]!;
    const b = parsed[i + 1]!;
    if (contactTimeMs >= a.t && contactTimeMs <= b.t) {
      if (a.t === b.t) return [a.lon, a.lat];
      const fraction = (contactTimeMs - a.t) / (b.t - a.t);
      return [a.lon + (b.lon - a.lon) * fraction, a.lat + (b.lat - a.lat) * fraction];
    }
  }

  return null;
}

/**
 * Primary dispatch function: compute the array centre for a sensor contact.
 *
 * Resolution order:
 *   1. If `sensor.offset` is null/undefined/0 → return host position unchanged
 *   2. If `sensor.array_centre_mode` is null/undefined → return host position unchanged
 *   3. Otherwise dispatch on `sensor.array_centre_mode`:
 *      - PLAIN: computePlainOffset()
 *      - WORM: backtrackAlongTrack()
 *      - MEASURED: interpolateMeasuredPosition(), falling back to PLAIN when null
 *
 * When mode is PLAIN or MEASURED fallback and `courseDeg` is null, the host
 * position is returned unchanged (no heading available to backtrack along).
 *
 * @param hostPosition Interpolated vessel position [lon, lat]
 * @param courseDeg Vessel course at contact time (degrees) or null
 * @param sensor Parent SensorData (provides offset, mode, measured_positions)
 * @param contactTimeMs Contact timestamp (epoch ms)
 * @param trackCoordinates Track geometry coordinates
 * @param trackPositions Track positions with timestamps
 * @returns Array centre [lon, lat]
 */
export function computeArrayCentre(
  hostPosition: [number, number],
  courseDeg: number | null,
  sensor: SensorData,
  contactTimeMs: number,
  trackCoordinates: [number, number][],
  trackPositions: Array<{ time: string }>,
): [number, number] {
  const offset = sensor.offset;
  const mode = sensor.array_centre_mode;

  if (offset === null || offset === undefined || offset <= 0) return hostPosition;
  if (mode === null || mode === undefined) return hostPosition;

  if (mode === 'PLAIN') {
    if (courseDeg === null) return hostPosition;
    return computePlainOffset(hostPosition, courseDeg, offset);
  }

  if (mode === 'WORM') {
    if (trackCoordinates.length === 0) return hostPosition;
    return backtrackAlongTrack(trackCoordinates, trackPositions, contactTimeMs, offset);
  }

  if (mode === 'MEASURED') {
    const measured = sensor.measured_positions;
    if (measured && measured.length > 0) {
      const interp = interpolateMeasuredPosition(measured, contactTimeMs);
      if (interp !== null) return interp;
    }
    // Fallback to PLAIN
    if (courseDeg === null) return hostPosition;
    return computePlainOffset(hostPosition, courseDeg, offset);
  }

  // Unknown mode — behave as if no offset applies
  return hostPosition;
}

/**
 * Convenience helper to resolve the course at a contact timestamp.
 * Wraps `interpolateTrackCourse` for symmetry with the Python side.
 */
export function resolveCourse(
  positions: Array<{ time: string; course?: number }>,
  contactTimeMs: number,
): number | null {
  return interpolateTrackCourse(positions, contactTimeMs);
}

File: services/calc/debrief_calc/tools/sensor/array_offset.py
"""Array offset calculations for towed-array sensors.

Three array centre modes determine where a sensor's bearing lines originate:
  - PLAIN: backtrack along the vessel's heading by the offset distance
  - WORM: walk backwards along the vessel's track path by the offset distance
  - MEASURED: interpolate from the sensor's measured position time-series,
              with fallback to PLAIN when the contact timestamp falls
              outside the measured range

All calculations are pure functions with identical behaviour to the TypeScript
implementation in shared/components/src/MapView/array-offset.ts. Parity is
verified by cross-language golden tests.
"""

from __future__ import annotations

import math
from datetime import datetime
from typing import Any

# Mean Earth radius in metres (matches TS haversineDistanceMetres)
EARTH_RADIUS_METRES = 6_371_000.0


# ── Time helpers ────────────────────────────────────────────────────


def _parse_iso_to_ms(time_iso: str) -> int:
    """Parse an ISO-8601 timestamp string to epoch milliseconds.

    Accepts trailing "Z" (UTC) and timezone offsets.
    """
    # Python's datetime.fromisoformat accepts +00:00 but not 'Z' before 3.11.
    # On 3.11+ it accepts 'Z' directly. Fallback for older paths:
    if time_iso.endswith("Z"):
        try:
            dt = datetime.fromisoformat(time_iso.replace("Z", "+00:00"))
        except ValueError as exc:
            raise ValueError(f"Invalid ISO-8601 timestamp: {time_iso!r}") from exc
    else:
        try:
            dt = datetime.fromisoformat(time_iso)
        except ValueError as exc:
            raise ValueError(f"Invalid ISO-8601 timestamp: {time_iso!r}") from exc
    return int(dt.timestamp() * 1000)


# ── Geo primitives ──────────────────────────────────────────────────


def haversine_distance_metres(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """Geodesic distance between two points in metres.

    Uses the haversine formula with the mean Earth radius (6371000m).

    Args:
        lon1: Longitude of point 1 (degrees).
        lat1: Latitude of point 1 (degrees).
        lon2: Longitude of point 2 (degrees).
        lat2: Latitude of point 2 (degrees).

    Returns:
        Distance in metres.
    """
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    d_phi = math.radians(lat2 - lat1)
    d_lambda = math.radians(lon2 - lon1)

    a = math.sin(d_phi / 2) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(d_lambda / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return EARTH_RADIUS_METRES * c


def _geodesic_destination(
    lon: float, lat: float, bearing_deg: float, distance_metres: float
) -> tuple[float, float]:
    """Destination point from (lon, lat) at given bearing and distance.

    Mirrors the TypeScript `geodesicDestination()` function from sensor-utils.ts
    (spherical Earth, Earth radius 6371000m).

    Returns:
        (lon, lat) of the destination in degrees.
    """
    lat1 = math.radians(lat)
    lon1 = math.radians(lon)
    brng = math.radians(bearing_deg)
    d = distance_metres / EARTH_RADIUS_METRES

    sin_d = math.sin(d)
    cos_d = math.cos(d)
    sin_lat1 = math.sin(lat1)
    cos_lat1 = math.cos(lat1)

    lat2 = math.asin(sin_lat1 * cos_d + cos_lat1 * sin_d * math.cos(brng))
    lon2 = lon1 + math.atan2(
        math.sin(brng) * sin_d * cos_lat1,
        cos_d - sin_lat1 * math.sin(lat2),
    )

    return (math.degrees(lon2), math.degrees(lat2))


# ── Mode algorithms ─────────────────────────────────────────────────


def compute_plain_offset(
    host_position: tuple[float, float],
    course_deg: float,
    offset_metres: float,
) -> tuple[float, float]:
    """PLAIN mode: backtrack from the vessel's position along its heading.

    The reverse bearing is ``(course_deg + 180) mod 360``. The array centre is
    the geodesic destination from the vessel at that reverse bearing, at the
    offset distance.
    """
    if offset_metres <= 0:
        return host_position
    reverse_bearing = ((course_deg + 180.0) % 360.0 + 360.0) % 360.0
    return _geodesic_destination(host_position[0], host_position[1], reverse_bearing, offset_metres)


def _interpolate_track_position(
    track_coordinates: list[list[float]],
    track_positions: list[dict[str, Any]],
    contact_time_ms: int,
) -> tuple[float, float] | None:
    """Linear interpolation of the track coordinate at a given timestamp.

    Mirrors the TypeScript `interpolateTrackPosition()`.
    Returns None when the contact time lies outside the track's time range.
    """
    if not track_coordinates or not track_positions:
        return None
    if len(track_coordinates) != len(track_positions):
        return None

    timestamps = [_parse_iso_to_ms(p["time"]) for p in track_positions]

    if contact_time_ms < timestamps[0] or contact_time_ms > timestamps[-1]:
        return None

    # Find bracketing pair
    for i in range(len(timestamps) - 1):
        t0 = timestamps[i]
        t1 = timestamps[i + 1]
        if t0 <= contact_time_ms <= t1:
            if t1 == t0:
                coord = track_coordinates[i]
                return (float(coord[0]), float(coord[1]))
            fraction = (contact_time_ms - t0) / (t1 - t0)
            lon0, lat0 = track_coordinates[i][0], track_coordinates[i][1]
            lon1, lat1 = track_coordinates[i + 1][0], track_coordinates[i + 1][1]
            return (
                lon0 + (lon1 - lon0) * fraction,
                lat0 + (lat1 - lat0) * fraction,
            )

    coord = track_coordinates[-1]
    return (float(coord[0]), float(coord[1]))


def backtrack_along_track(
    track_coordinates: list[list[float]],
    track_positions: list[dict[str, Any]],
    contact_time_iso: str,
    offset_metres: float,
) -> tuple[float, float]:
    """WORM mode: walk backward along the vessel's track path.

    Starts at the interpolated vessel position at the contact time, then walks
    backward through track segments accumulating haversine distances until the
    offset distance is reached. If the track is exhausted before reaching the
    offset, the array centre is placed at the earliest available track point.

    Args:
        track_coordinates: Track geometry coordinates as a list of [lon, lat] pairs.
        track_positions: Track positions, each with an ISO-8601 ``time`` field.
        contact_time_iso: Contact timestamp (ISO-8601 string).
        offset_metres: Distance to walk backward along the track path.

    Returns:
        Array centre (lon, lat) on the track path.
    """
    if not track_coordinates:
        raise ValueError("backtrack_along_track: empty track coordinates")
    if len(track_coordinates) == 1 or offset_metres <= 0:
        first = track_coordinates[0]
        return (float(first[0]), float(first[1]))

    contact_time_ms = _parse_iso_to_ms(contact_time_iso)
    timestamps = [_parse_iso_to_ms(p["time"]) for p in track_positions]

    start_point = _interpolate_track_position(track_coordinates, track_positions, contact_time_ms)
    if start_point is None:
        first = track_coordinates[0]
        start_point = (float(first[0]), float(first[1]))

    # Determine starting segment index
    if contact_time_ms >= timestamps[-1]:
        start_idx = len(track_coordinates) - 1
    elif contact_time_ms <= timestamps[0]:
        first = track_coordinates[0]
        return (float(first[0]), float(first[1]))
    else:
        # Find largest index i where timestamps[i] <= contact_time_ms
        start_idx = 0
        for i, t in enumerate(timestamps):
            if t <= contact_time_ms:
                start_idx = i
            else:
                break

    remaining = offset_metres
    current_point = (start_point[0], start_point[1])

    for i in range(start_idx, 0, -1):
        prev_point = track_coordinates[i - 1]
        seg_len = haversine_distance_metres(
            current_point[0], current_point[1], prev_point[0], prev_point[1]
        )

        if seg_len == 0:
            current_point = (float(prev_point[0]), float(prev_point[1]))
            continue

        if remaining <= seg_len:
            fraction = remaining / seg_len
            return (
                current_point[0] + (prev_point[0] - current_point[0]) * fraction,
                current_point[1] + (prev_point[1] - current_point[1]) * fraction,
            )

        remaining -= seg_len
        current_point = (float(prev_point[0]), float(prev_point[1]))

    # Track exhausted: return the earliest point on the track
    first = track_coordinates[0]
    return (float(first[0]), float(first[1]))


def interpolate_measured_position(
    measured_positions: list[dict[str, Any]],
    contact_time_iso: str,
) -> tuple[float, float] | None:
    """MEASURED mode: interpolate from the sensor's measured position time-series.

    If the contact time is outside the measured range, returns ``None`` so the
    caller can fall back to PLAIN mode (FR-004).

    Measured positions are sorted by time before lookup to handle unordered input.
    """
    if not measured_positions:
        return None

    parsed: list[tuple[int, float, float]] = []
    for p in measured_positions:
        try:
            t_ms = _parse_iso_to_ms(p["time"])
        except (ValueError, KeyError):
            continue
        loc = p.get("location")
        if not loc or len(loc) < 2:
            continue
        parsed.append((t_ms, float(loc[0]), float(loc[1])))

    if not parsed:
        return None

    parsed.sort(key=lambda x: x[0])

    contact_time_ms = _parse_iso_to_ms(contact_time_iso)
    first_t = parsed[0][0]
    last_t = parsed[-1][0]

    if contact_time_ms < first_t or contact_time_ms > last_t:
        return None

    if contact_time_ms == first_t:
        return (parsed[0][1], parsed[0][2])
    if contact_time_ms == last_t:
        return (parsed[-1][1], parsed[-1][2])

    for i in range(len(parsed) - 1):
        t0, lon0, lat0 = parsed[i]
        t1, lon1, lat1 = parsed[i + 1]
        if t0 <= contact_time_ms <= t1:
            if t1 == t0:
                return (lon0, lat0)
            fraction = (contact_time_ms - t0) / (t1 - t0)
            return (
                lon0 + (lon1 - lon0) * fraction,
                lat0 + (lat1 - lat0) * fraction,
            )

    return None


# ── Course interpolation helper ─────────────────────────────────────


def interpolate_track_course(
    track_positions: list[dict[str, Any]],
    contact_time_iso: str,
) -> float | None:
    """Resolve the vessel course at a contact timestamp.

    Uses nearest-neighbour lookup among the track positions (matches the
    TypeScript `interpolateTrackCourse` behaviour). Returns ``None`` when the
    contact time is outside the track's time range or no course is available.
    """
    if not track_positions:
        return None
    timestamps = [_parse_iso_to_ms(p["time"]) for p in track_positions]
    contact_time_ms = _parse_iso_to_ms(contact_time_iso)

    if contact_time_ms < timestamps[0] or contact_time_ms > timestamps[-1]:
        return None

    # Nearest-neighbour by absolute timestamp difference
    best_idx = 0
    best_diff = abs(timestamps[0] - contact_time_ms)
    for i, t in enumerate(timestamps):
        diff = abs(t - contact_time_ms)
        if diff < best_diff:
            best_diff = diff
            best_idx = i

    course = track_positions[best_idx].get("course")
    if course is None:
        return None
    return float(course)


# ── Dispatcher ──────────────────────────────────────────────────────


def compute_array_centre(
    host_position: tuple[float, float],
    course_deg: float | None,
    offset_metres: float | None,
    array_centre_mode: str | None,
    measured_positions: list[dict[str, Any]] | None,
    contact_time_iso: str,
    track_coordinates: list[list[float]],
    track_positions: list[dict[str, Any]],
) -> tuple[float, float]:
    """Primary dispatch: compute the array centre for a sensor contact.

    Resolution order:
      1. If ``offset_metres`` is None/0 → return host position unchanged
      2. If ``array_centre_mode`` is None → return host position unchanged
      3. Otherwise dispatch on mode:
         - PLAIN: compute_plain_offset()
         - WORM: backtrack_along_track()
         - MEASURED: interpolate_measured_position(); on None, fall back to PLAIN

    Args:
        host_position: Interpolated vessel position (lon, lat).
        course_deg: Vessel course at contact time (degrees) or None.
        offset_metres: Sensor offset distance in metres.
        array_centre_mode: "PLAIN", "WORM", "MEASURED", or None.
        measured_positions: Sensor's measured position list (for MEASURED mode).
        contact_time_iso: Contact timestamp (ISO-8601 string).
        track_coordinates: Vessel track coordinates.
        track_positions: Vessel track positions (with timestamps).

    Returns:
        Array centre (lon, lat).
    """
    if offset_metres is None or offset_metres <= 0:
        return host_position
    if array_centre_mode is None:
        return host_position

    if array_centre_mode == "PLAIN":
        if course_deg is None:
            return host_position
        return compute_plain_offset(host_position, course_deg, offset_metres)

    if array_centre_mode == "WORM":
        if not track_coordinates:
            return host_position
        return backtrack_along_track(
            track_coordinates, track_positions, contact_time_iso, offset_metres
        )

    if array_centre_mode == "MEASURED":
        if measured_positions:
            interp = interpolate_measured_position(measured_positions, contact_time_iso)
            if interp is not None:
                return interp
        # Fallback to PLAIN
        if course_deg is None:
            return host_position
        return compute_plain_offset(host_position, course_deg, offset_metres)

    # Unknown mode — behave as if no offset applies
    return host_position
